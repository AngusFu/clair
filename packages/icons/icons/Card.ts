export const Card = `
<svg viewBox="0 0 28 28">
  <g fill="none" fill-rule="evenodd">
    <path d="M3.8 3h20.4A2.8 2.8 0 0127 5.8v16.4a2.8 2.8 0 01-2.8 2.8H3.8A2.8 2.8 0 011 22.2V5.8A2.8 2.8 0 013.8 3zM0 9h28" stroke="currentColor"/>
    <path d="M18 19h4" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round"/>
  </g>
</svg>`;
