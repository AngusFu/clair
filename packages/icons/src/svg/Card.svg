<?xml version="1.0" encoding="UTF-8"?>
<svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 56.2 (81672) - https://sketch.com -->
    <title>Card</title>
    <desc>Created with Sketch.</desc>
    <g id="Card" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g>
            <rect id="蒙版" opacity="0.200000003" x="0" y="0" width="28" height="28"></rect>
            <path d="M3.8,3 L24.2,3 C25.7463973,3 27,4.2536027 27,5.8 L27,22.2 C27,23.7463973 25.7463973,25 24.2,25 L3.8,25 C2.2536027,25 1,23.7463973 1,22.2 L1,5.8 C1,4.2536027 2.2536027,3 3.8,3 Z" id="矩形" stroke="#6E6E6E" stroke-width="2"></path>
            <path d="M0,9 L28,9" id="路径-54" stroke="#6E6E6E" stroke-width="2"></path>
            <path d="M18,19 L22,19" id="路径-55" stroke="#6E6E6E" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"></path>
        </g>
    </g>
</svg>