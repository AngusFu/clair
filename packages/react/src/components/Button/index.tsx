import classNames from "classnames";
import React from "react";
import PropTypes from "prop-types";

const types = [
  "primary",
  "default",
  "warning",
  "danger",
  "success",
  "ghost"
] as const;
export type Type = typeof types[number];
const PType = PropTypes.oneOf([...types]);

const sizes = ["large", "normal", "small"] as const;
export type Size = typeof sizes[number];
const PSize = PropTypes.oneOf([...sizes]);

const htmlTypes = ["button", "reset", "submit"] as const;
export type HtmlType = typeof htmlTypes[number];
const PHtmlType = PropTypes.oneOf([...htmlTypes]);

// handle "detail" property of custom event
type ButtonEvent = React.MouseEvent<HTMLButtonElement> &
  Partial<Pick<CustomEvent, "detail">>;
type EventHandler = React.EventHandler<ButtonEvent>;

export interface ButtonProps {
  size?: Size;
  type?: Type;
  htmlType?: HtmlType;
  disabled?: boolean;
  loading?: boolean;
  block?: boolean;
  className?: string;
  style?: React.CSSProperties;
  children?: React.ReactNode;
  onClick?: EventHandler;
  onSubmit?: EventHandler;
  onReset?: EventHandler;
  forwardRef?: React.RefObject<HTMLButtonElement>;
}

export const Button: React.FC<ButtonProps> = props => {
  const {
    size = "normal",
    type = "default",
    htmlType = "button",
    disabled = false,
    loading = false,
    block = false,
    children,
    forwardRef,
    className,
    ...rest
  } = props;

  const className2Use: string = classNames("c-button", className, {
    [`c-button--${type}`]: types.some(t => t === type),
    [`c-button--${size}`]: sizes.some(s => s === size),
    "c-button--block": block,
    "c-button--loading": loading
  });
  return (
    <button
      ref={forwardRef}
      className={className2Use}
      disabled={disabled}
      type={htmlType}
      {...rest}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  size: PSize,
  type: PType,
  htmlType: PHtmlType,
  disabled: PropTypes.bool,
  block: PropTypes.bool,
  loading: PropTypes.bool,
  className: PropTypes.string,
  style: PropTypes.object,
  onClick: PropTypes.func,
  onSubmit: PropTypes.func,
  onReset: PropTypes.func,
  children: PropTypes.node,
  forwardRef: PropTypes.shape({
    current: PropTypes.any
  })
} as PropTypes.ValidationMap<ButtonProps>;

Button.displayName = "Button";
