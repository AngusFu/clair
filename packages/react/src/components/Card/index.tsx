import classNames from "classnames";
import React from "react";
import PropTypes from "prop-types";

const sizes = ["normal", "small"] as const;
export type Size = typeof sizes[number];
const PSize = PropTypes.oneOf([...sizes]);

export interface CardProps {
  title?: React.ReactNode;
  size?: Size;
  raised?: boolean;
  actions?: React.ReactNode;
  cover?: React.ReactNode;
  children?: React.ReactNode;
}

export const Card: React.FC<CardProps> = props => {
  const {
    title,
    size = "normal",
    raised = false,
    actions,
    cover,
    children
  } = props;
  const className: string = classNames("c-card", {
    [`c-card--${size}`]: sizes.includes(size) && size !== "normal",
    "c-card--raised": raised
  });
  const Cover = cover ? <div className="c-card__cover">{cover}</div> : null;
  const Title = title ? <h3 className="c-card__title">{title}</h3> : null;
  const Actions = actions ? (
    <div className="c-card__action">{actions}</div>
  ) : null;
  const Header =
    Title || Actions ? (
      <div className="c-card__header">
        {Title}
        {Actions}
      </div>
    ) : null;
  const Body = children ? <div className="c-card__body">{children}</div> : null;
  return (
    <div className={className}>
      {Cover}
      {Header}
      {Body}
    </div>
  );
};

Card.propTypes = {
  title: PropTypes.node,
  size: PSize,
  raised: PropTypes.bool,
  actions: PropTypes.node,
  cover: PropTypes.node,
  children: PropTypes.node
};

Card.displayName = "Card";
