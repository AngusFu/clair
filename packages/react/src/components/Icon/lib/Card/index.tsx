import { IconContainer, IconProps } from "@components/Icon/lib/Container";
import { Card } from "@clair/icons";
import { getStyleMergedComponent } from "@src/utils";

export const IconCard = getStyleMergedComponent<IconProps>({
  template: Card
})(IconContainer);
