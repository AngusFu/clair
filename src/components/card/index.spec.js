import { mount } from '@vue/test-utils'
import Card from './index.vue'
import Button from '../button/index.vue'

it('should render card correctly', () => {
  const wrapper = mount(Card, {
    stubs: {
      'c-button': Button
    },
    slots: {
      title: '<h3>卡片标题</h3>',
      default: '<div>这里是卡片介绍</div>',
      actions: [
        '<c-button flat primary icon="heart">加入收藏</c-button>',
        '<c-button flat icon="download">下载</c-button>'
      ]
    }
  })
  expect(wrapper).toMatchSnapshot()
})

it('should render horizontal card correctly', () => {
  const wrapper = mount(Card, {
    propsData: { horizontal: true },
    stubs: {
      'c-button': Button
    },
    slots: {
      default: '<div>云中温度低于0℃的许多小云滴在冰晶上互相碰撞凝结形成雪珠，小雪珠是由许多细白的冰粒聚集而成的。</div>',
      media: '<img src="https://p0.ssl.qhimg.com/dm/100_100_90/t019549f24fd684aeb8.jpg" />'
    }
  })
  expect(wrapper).toMatchSnapshot()
})
