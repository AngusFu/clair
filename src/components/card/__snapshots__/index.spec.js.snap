// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render card correctly 1`] = `
<div class="c-card">
  <div class="c-card__title">
    <h3>卡片标题</h3>
  </div>
  <!---->
  <div class="c-card__body">
    <div>这里是卡片介绍</div>
  </div>
  <div class="c-card__actions"><button type="button" class="c-button c-button--primary c-button--flat"><span class="c-icon">​<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-heart" style="vertical-align: middle;"><path d="M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z"></path></svg></span><span>加入收藏</span></button><button type="button" class="c-button c-button--flat"><span class="c-icon">​<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-download" style="vertical-align: middle;"><path d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"></path><polyline points="7 10 12 15 17 10"></polyline><line x1="12" y1="15" x2="12" y2="3"></line></svg></span><span>下载</span></button></div>
</div>
`;

exports[`should render horizontal card correctly 1`] = `
<div class="c-card is-horizontal">
  <!---->
  <div class="c-card__media"><img src="https://p0.ssl.qhimg.com/dm/100_100_90/t019549f24fd684aeb8.jpg"></div>
  <div class="c-card__body">
    <div>云中温度低于0℃的许多小云滴在冰晶上互相碰撞凝结形成雪珠，小雪珠是由许多细白的冰粒聚集而成的。</div>
  </div>
  <!---->
</div>
`;
